streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -267,9 +271,25 @@ private void maybeClean() {
                 for (File dir : stateDirs) {
                     try {
                         Integer id = Integer.parseInt(dir.getName());
-                        if (!tasks.containsKey(id)) {
-                            log.info("Deleting obsolete state directory {} after delayed {} ms.", dir.getAbsolutePath(), cleanTimeMs);
-                            Utils.delete(dir);
+
+                        // try to acquire the exclusive lock on the state directory
+                        FileLock directoryLock = null;
+                        try {
+                            directoryLock = ProcessorStateManager.lockStateDirectory(dir);
+                            if (directoryLock != null) {
+                                log.info("Deleting obsolete state directory {} after delayed {} ms.", dir.getAbsolutePath(), cleanTimeMs);
+                                Utils.delete(dir);
+                            }
+                        } catch (IOException e) {
+                            log.error("Failed to lock the state directory", e);
```

Comment:
This is an unexpected error. I will change the error text to make it clear. 
When tryLock() cannot acquire a lock, OverlappingFileLockException is thrown instead of IOException.